### PR TITLE
Pheno browser pagination

### PR DESCRIPTION
**Pull request by IlinaGergova**
## Background

Need to make phenotype browser table using pagination to be able to sort the table correctly.

## Aim

To make it.

## Implementation

Remove gpf table and make new table instead (using grid). Initially loaded rows are 25 and when scrolling new 25 rows load. Sorting is happening in the backend. To get measures we pass selected instrument, search text,  last loaded page number, current dataset id, sort by (optional) and order by (optional). When changing instrument or searching the sorting and the page counter reset.

UPDATE: Table is not supporting pagination. It loads 1000 measures initially. Above the table measure counter is added similar to Genotype browser/ Gene browser. Sorting is made locally.

### CODE REVIEW

**Review by joankosev — APPROVED** [review]

File: src/app/pheno-browser-table/pheno-browser-table.component.html [hunk 1]
```diff
@@ -1,144 +1,197 @@
...
+  [ngStyle]="{
+    'grid-template-columns': 'repeat(' + columnsCount + ', 4fr)'
+  }">
+  <div
+    class="grid-header"
+    (click)="sort('instrumentName', 'instrumentName' === sortBy && orderBy === 'desc' ? 'asc' : 'desc')">
+    <div class="single-sorting-button">
+      <span>Instrument</span>
+      <gpf-sorting-buttons
+        *ngIf="true"
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.html:18]: Remove?

File: src/app/pheno-browser-table/pheno-browser-table.component.css [hunk 2]
```diff
@@ -16,3 +9,79 @@ gpf-table {
...
   height: 100%;
 }
+
+.grid-container {
+  padding-top: 10px;
+  min-width: 1200px;
+  display: grid;
+}
+
+.grid-header {
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.css:19]: Rename grid-header to grid-header-cell

File: src/app/pheno-browser-table/pheno-browser-table.component.html [hunk 3]
```diff
@@ -1,144 +1,197 @@
...
+    </div>
+  </div>
+
+  <div
+    class="grid-header"
+    (click)="sort('measureName', 'measureName' === sortBy && orderBy === 'desc' ? 'asc' : 'desc')">
+    <div class="single-sorting-button">
+      <span>Proband Pheno Measure</span>
+      <gpf-sorting-buttons
+        *ngIf="true"
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.html:32]: Remove?

File: src/app/pheno-browser-table/pheno-browser-table.component.html [hunk 4]
```diff
@@ -1,144 +1,197 @@
...
+  </ng-container>
+
+  <div
+    *ngIf="measures.measures !== null && !measures.measures.length"
+    id="nothing-found-row"
+    [ngStyle]="{ 'grid-column': '1 /' + (columnsCount + 1) }"
+    >Nothing found</div
+  >
+
+  <ng-container *ngFor="let measure of measures.measures" class="grid-row">
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.html:109]: Remove class="grid-row"

File: src/app/pheno-browser-table/pheno-browser-table.component.html [hunk 5]
```diff
@@ -1,144 +1,197 @@
...
+        <img *ngIf="measure.figureDistributionSmall" class="table-chart" [src]="measure.figureDistributionSmall" />
+      </a>
+    </div>
+
+    <div class="grid-cell measure-type-cell">
+      <div>
+        <div>{{ measure.measureType }}</div>
+        <ng-container *ngIf="measure.measureType === 'continuous'">
+          <div
+            title="[{{
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.html:129]: Remove title

File: src/app/pheno-browser-table/pheno-browser-table.component.ts [hunk 6]
```diff
@@ -1,58 +1,119 @@
...
+
+    this.sortTable(this.measures.measures, this.orderBy, this.sortBy);
+  }
+
+  public sortTable(data: PhenoMeasure[], orderBy: string, sortBy: string): void {
+    data.sort((m1, m2) => {
+      let a: string | number;
+      let b: string | number;
+      if (sortBy.includes('pvalueRegressionMale')) {
+        const arr = sortBy.split('.');
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.ts:70]: Change `const arr = sortBy.split('.');` to `const regId = sortBy.split('.')[0];`
And down use `regId`

File: src/app/pheno-browser-table/pheno-browser-table.component.ts [hunk 7]
```diff
@@ -1,58 +1,119 @@
...
+        a = m1.regressions.getReg(arr[0]).pvalueRegressionMale;
+        b = m2.regressions.getReg(arr[0]).pvalueRegressionMale;
+      } else if (sortBy.includes('pvalueRegressionFemale')) {
+        const arr = sortBy.split('.');
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.ts:74]: Change const arr = sortBy.split('.'); to const regId = sortBy.split('.')[0];
And down use regId

File: src/app/pheno-browser-table/pheno-browser-table.component.ts [hunk 8]
```diff
@@ -1,58 +1,119 @@
...
+      return a - b;
     }
+    if (isString(a) && isString(b)) {
+      return a.localeCompare(b);
+    }
+    return 0;
+  }
 
-    return leftVal.localeCompare(rightVal);
+  private resetSortButtons(): void {
```
Comment (IvoStefanov) [inline src/app/pheno-browser-table/pheno-browser-table.component.ts:109]: Change `resetSortButtons` to `resetCurrentSortButton`

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 9]
```diff
@@ -22,21 +21,20 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
-
   public selectedDataset: Dataset;
-
   public input$ = new ReplaySubject<string>(1);
-
   @ViewChild('searchBox') public searchBox: ElementRef;
-
   public imgPathPrefix = environment.imgPathPrefix;
 
+  private getPageSubscription: Subscription = new Subscription();
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:35]: Remove?

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 10]
```diff
@@ -59,6 +57,7 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
       this.initMeasuresToShow(this.selectedDataset.id);
     });
 
+    this.getPageSubscription?.unsubscribe();
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:60]: Remove?

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 11]
```diff
@@ -70,51 +69,18 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
-        }
-        return this.measuresToShow;
-      }),
-      share()
-    ).subscribe();
+        return this.phenoBrowserService.getMeasuresInfo(datasetId);
+      })
+    ).subscribe(phenoMeasures => {
+      this.measuresToShow = phenoMeasures;
+      this.measuresToShow?.clear();
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:80]: Remove?

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 12]
```diff
@@ -125,12 +91,60 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
+    this.location.replaceState(url);
+  }
+
   private initInstruments(datasetId: string): void {
     this.instruments = this.phenoBrowserService.getInstruments(datasetId);
   }
 
   public emitInstrument(instrument: PhenoInstrument): void {
     this.selectedInstrument$.next(instrument);
+    this.measuresToShow?.clear();
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:121]: Remove?

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 13]
```diff
@@ -125,12 +91,60 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
+    this.measuresChangeTick++;
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:122]: Remove.

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 14]
```diff
@@ -125,12 +91,60 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
+
+  private updateTable(): void {
+    this.getPageSubscription?.unsubscribe();
+    this.getPageSubscription =
+      this.phenoBrowserService.getMeasures(
+        this.selectedDataset.id,
+        this.selectedInstrument$.value,
+        (this.searchBox.nativeElement as HTMLInputElement).value
+      ).subscribe(res => {
+        if (!res.length && this.measuresToShow.measures) {
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:133]: Remove `&& this.measuresToShow.measures`

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 15]
```diff
@@ -125,12 +91,60 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
     this.instruments = this.phenoBrowserService.getInstruments(datasetId);
   }
 
   public emitInstrument(instrument: PhenoInstrument): void {
     this.selectedInstrument$.next(instrument);
+    this.measuresToShow?.clear();
+    this.measuresChangeTick++;
+  }
+
+  private updateTable(): void {
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:125]: Call `this.measuresToShow?.clear();`

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 16]
```diff
@@ -162,6 +176,8 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
 
   public search(value: string): void {
     this.input$.next(value);
+    this.measuresToShow?.clear();
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:179]: Remove!

File: src/app/pheno-browser/pheno-browser.component.ts [hunk 17]
```diff
@@ -162,6 +176,8 @@ export class PhenoBrowserComponent implements OnInit, OnDestroy {
...
+    this.measuresChangeTick++;
```
Comment (IvoStefanov) [inline src/app/pheno-browser/pheno-browser.component.ts:180]: Remove!

**Pull request merged**